docs/src/apalache/fold.md

```diff
@@ -0,0 +1,213 @@
+
+  1. If `C` is empty, the resulting value is `b`
+  1. If `C` is nonempty, we select an element `x` from `C`. If `C` is a sequence, `x` is the head value, and if `C` is a set, `x` is an arbitrary member. Then, we compute `b2 = Op(b, x)` and take `C2 = C without x` (`Tail(C)` or `C \ {x}` respectively). We continue the process, by computing the fold of `Op` over `C2` starting with `b2`.
+
+Note that this is the definition of a _left fold_ in the literature. Apalache does not implement a right fold.
+
+For example, if `C` is the sequence `<<x,y,z>>`, the result is equal to `Op( Op( Op(b, x), y), z)`. If `C = {x,y}`, the result is either `Op( Op(b, x), y)` or `Op( Op(b, y), x)`. Because the order of elements selected is arbitrary in the case of sets, users should be careful, as the result is only uniquely defined in the case that the operator is both associative (`Op(Op(a,b),c) = Op(a,Op(b,c))`) and commutative (`Op(a,b) = Op(b,a)`).
+
+### Using fold operators in Apalache
+
+The syntax of the fold operators is as follows:
+
+```
+\* @type: ( (a, b) => a, a, Set(b) ) => a
+FoldSet( operator, base, set )
+
+\* @type: ( (a, b) => a, a, Seq(b) ) => a
+FoldSeq( operator, base, seq )
+```
+
+As shown by the type signature, Apalache permits a very general form of folding, where the types of the collection elements and the type of the base element/return-type of the operator do not have to match. Again, we urge users to exercise caution when using `FoldSet` with an operator, for which the types `a` and `b` are different, as such operators cannot be commutative or associative, and therefore the result is not guaranteed to be unique and predictable. 
+
+The other component of note is `operator`, the _name_ (not definition) of some binary operator, which is available in this context. The following are examples of valid uses of folds:
+
+```
+PlusOne(p,q) == p + q + 1
+X == FoldSet( PlusOne, 0, {1,2,3} ) \* X = 9 
+
+X == LET Count(p,q) == p + 1 IN FoldSeq( Count, 0, <<1,2,3>> ) \* X = 3
+```
+
+while these next examples are considered invalid:
+
+```
+\* LAMBDAS in folds are not supported
+\* Use a local LET-IN operator instead 
+X == FoldSet( LAMBDA p,q: p + q, 0, {1,2,3} ) 
+
+\* Built-in operators cannot be called by name
+\* Define a LET-IN operator Plus(p,q) == p + q instead
+X == FoldSet( + , 0, {1,2,3} )
+```
+
+Local LET definitions can also be used as closures:
+
+```
+A(x) == LET PlusX(p,q) == p + q + x IN FoldSeq( PlusX, 0, <<1,2,3>> )
+
+X == A(1) \* X = 9
+```
+
+### Folding VS recursion
+
+While TLA+ allows users to write arbitrary recursive operators, they are, in our experience, mostly used to implement collection traversals. Consider the following implementations of a `Max` operator, which returns the largest element of a sequence:
+
+```
+CONSTANT negInf \* Max(<<>>) = -inf, but TLA+ has no literal for ±inf, so we use it as a constant
+
+RECURSIVE MaxRec(_)
+MaxRec(seq) == IF seq = <<>>
```

Comment:
Yes, but we're explicitly comparing fold to recursion here.

Reply:
Right. But it would be good to have pros and cons of the both approaches. You can write this recursive operator without `CHOOSE`, but it is not how TLC users would do it.